src/Nethermind/Nethermind.Network.Discovery/Kademlia/KBucketTree.cs

```diff
@@ -0,0 +1,130 @@
+// SPDX-FileCopyrightText: 2024 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using Nethermind.Core.Crypto;
+
+namespace Nethermind.Network.Discovery.Kademlia;
+
+public class KBucketTree<TNode, TContentKey> where TNode : notnull
+{
+    private class TreeNode
+    {
+        public KBucket<TNode> Bucket { get; }
+        public TreeNode? Left { get; set; }
+        public TreeNode? Right { get; set; }
+        public int Depth { get; }
+
+        public TreeNode(int depth, int k)
+        {
+            Bucket = new KBucket<TNode>(k);
+            Depth = depth;
+        }
+    }
+
+    private readonly TreeNode _root;
+    private readonly int _k;
+    private readonly int _maxDepth;
+    
+    private readonly INodeHashProvider<TNode, TContentKey> _nodeHashProvider;
+
+    public KBucketTree(int k, int maxDepth, INodeHashProvider<TNode, TContentKey> nodeHashProvider)
+    {
+        _k = k;
+        _maxDepth = maxDepth;
+        _nodeHashProvider = nodeHashProvider;
+        _root = new TreeNode(0, k);
+    }
+
+    public bool TryAddOrRefresh(TNode node, out TNode? toRefresh)
```

Comment:
Please read the kademlia paper carefully and check other implementation also.
The key to the tree is the XOR of the nodehash againts current node hash. 
The tree does not have a fixed maxdepth. It started with only one tree node (one bucket) and when the bucket is full, it is splitted into two, distributing the nodes into two bucket. A bucket is only split when full if the current node hash can be contained within that bucket. There are other optimization such as splitting a bucket even if it does not contain the current node, but that is for later.

Reply:
Made a lot of changes now this is in line with the Kademlia paper. Start with one k bucket and bucket is full split it. And mainly the XOR metric for distance.